src/crypto.h

```diff
@@ -43,6 +45,34 @@ struct PublicKey {
   int key_length;
 };
 
+class MultiPartSHA512Hasher {
+ public:
+  MultiPartSHA512Hasher() { crypto_hash_sha512_init(&state_); }
+  void update(const unsigned char *part, int64_t size) { crypto_hash_sha512_update(&state_, part, size); }
+  std::string getHexDigest() {
+    std::string sha512_hash(crypto_hash_sha512_BYTES, '\0');
+    crypto_hash_sha512_final(&state_, (unsigned char *)sha512_hash.c_str());
```

Comment:
Safe... except the C++ spec explicitly disallows modification of the result of c_str() "Writing to the character array accessed through c_str() is undefined behavior."

We should not be entering into the underground, murky, dangerous world of undefined behaviour lightly :)